File: DSA/1. Hungry Heroes/Part 3/Java/Judge.java
import java.util.*;

public class Judge {

    public void runTests() {

        String[][] testCasesS = {
            {"Beef", "2", "Chicken", "1", "Pork", "1", "Vegetable", "3", "Grain", "1"},
            {"Beef", "5", "Pork", "2", "Vegetable", "1", "Grain", "3"},
            {"Grain", "30"},
            {"Beef", "1", "Chicken", "1", "Pork", "1", "Vegetable", "1", "Grain", "1", "Fruit", "1"}
        };

        String[][][] testCasesP = {
            {
                {"Beef", "2", "Chicken", "1", "Pork", "2", "Vegetable", "0.5", "Grain", "1", "Fruit", "0.5"},
                {"Beef", "0.5", "Chicken", "1", "Pork", "0.5", "Vegetable", "2", "Grain", "1", "Fruit", "2"},
                {"Beef", "2", "Chicken", "2", "Pork", "0.5", "Vegetable", "1", "Grain", "0.5", "Fruit", "1"},
                {"Beef", "1", "Chicken", "2", "Pork", "2", "Vegetable", "0.5", "Grain", "1", "Fruit", "1"}
            },
            {
                {"Beef", "2", "Chicken", "1", "Pork", "2", "Vegetable", "0.5", "Grain", "1", "Fruit", "0.5"},
                {"Beef", "0.5", "Chicken", "1", "Pork", "0.5", "Vegetable", "2", "Grain", "1", "Fruit", "2"},
                {"Beef", "2", "Chicken", "2", "Pork", "0.5", "Vegetable", "1", "Grain", "0.5", "Fruit", "1"},
                {"Beef", "1", "Chicken", "2", "Pork", "2", "Vegetable", "0.5", "Grain", "1", "Fruit", "1"}
            },
            {
                {"Beef", "2", "Chicken", "1", "Pork", "2", "Vegetable", "0.5", "Grain", "1", "Fruit", "0.5"},
                {"Beef", "0.5", "Chicken", "1", "Pork", "0.5", "Vegetable", "2", "Grain", "1", "Fruit", "2"},
                {"Beef", "2", "Chicken", "2", "Pork", "0.5", "Vegetable", "1", "Grain", "0.5", "Fruit", "1"},
                {"Beef", "1", "Chicken", "2", "Pork", "2", "Vegetable", "0.5", "Grain", "1", "Fruit", "1"}
            },
            {
                {"Beef", "2", "Chicken", "1", "Pork", "2", "Vegetable", "0.5", "Grain", "1", "Fruit", "0.5"},
                {"Beef", "0.5", "Chicken", "1", "Pork", "0.5", "Vegetable", "2", "Grain", "1", "Fruit", "2"},
                {"Beef", "2", "Chicken", "2", "Pork", "0.5", "Vegetable", "1", "Grain", "0.5", "Fruit", "1"},
                {"Beef", "1", "Chicken", "2", "Pork", "2", "Vegetable", "0.5", "Grain", "1", "Fruit", "1"}
            }
        };

        int[][] testCasesR = {
            {2, 2, 2, 2},
            {2, 2, 2, 2},
            {10, 10, 10, 10},
            {4, 3, 2, 1}
        };

        runTests(testCasesS, testCasesP, testCasesR);
    }

    public void runTests(String[][] s, String[][][] p, int[][] r) {
        for (int i = 0; i < s.length; i++) {
            test(s[i], p[i], r[i], testCaseSolutionOutput(s[i], p[i], r[i]));
        }
    }

    private boolean test(String[] s, String[][] p, int[] r, boolean expected) {
        Solution sol = new Solution();
        boolean output = sol.distributeFood(s, p, r);
        boolean passed = output == expected;

        System.out.println("Test Case:");
        System.out.println("s: " + Arrays.toString(s));
        System.out.println("p: " + Arrays.deepToString(p));
        System.out.println("r: " + Arrays.toString(r));
        System.out.println("Expected: " + expected);
        System.out.println("Output: " + output);
        System.out.println(passed ? "Test PASSED" : "Test FAILED");
        System.out.println("-----------");

        return passed;
    }

    private boolean testCaseSolutionOutput(String[] s, String[][] p, int[] r) {

    }

    private boolean backtrack() {

    }

}

File: DSA/1. Hungry Heroes/Part 2 (Easy)/JavaScript/judge.js
const Solution = require('./solution');

class Judge {

    runTests() {
        const testCasesC = [
            ["Beef", "Beef", "Chicken", "Pork", "Vegetable", "Grain", "Vegetable", "Vegetable"],
            ["Fruit", "Beef", "Vegetable"],
            ["Beef", "Beef", "Beef"],
            ["Fruit", "Beef", "Vegetable"],
        ];

        const testCasesP = [
            ["Beef", "2", "Chicken", "1", "Pork", "2", "Vegetable", "0.5", "Grain", "1", "Fruit", "0.5"],
            ["Beef", "2", "Chicken", "1", "Pork", "2", "Vegetable", "0.5", "Grain", "1", "Fruit", "0.5"],
            ["Beef", "2", "Chicken", "1", "Pork", "2", "Vegetable", "0.5", "Grain", "1", "Fruit", "0.5"],
            ["Beef", "2", "Chicken", "1", "Pork", "2", "Vegetable", "0.5", "Grain", "1", "Fruit", "0.5"],
        ];

        const testCasesQ = [
            [[1, 3], [0, 4], [2, 6]],
            [[0, 2]],
            [[0, 2]],
            [[0, 1]],
        ];

        for (let i = 0; i < testCasesC.length; i++) {
            this.test(testCasesC[i], testCasesP[i], testCasesQ[i], this.testCaseSolutionOutput(testCasesC[i], testCasesP[i], testCasesQ[i]));
        }
    }

    test(c, p, q, expected) {
        const sol = new Solution();
        const output = sol.calculateSatisfaction(c, p, q);
        const passed = this.arrayEquals(output, expected);

        console.log("Test Case:");
        console.log("c:", JSON.stringify(c));
        console.log("p:", JSON.stringify(p));
        console.log("q:", JSON.stringify(q));
        console.log("Expected:", JSON.stringify(expected));
        console.log("Output:", JSON.stringify(output));
        console.log(passed ? "Test PASSED" : "Test FAILED");
        console.log("-----------");

        return passed;
    }

    arrayEquals(a, b) {
        return JSON.stringify(a) === JSON.stringify(b);
    }

    testCaseSolutionOutput(c, p, q) {
        const satisfactionMap = new Map();
        for (let i = 0; i < p.length; i += 2) {
            const food = p[i];
            const satisfaction = parseFloat(p[i + 1]);
            satisfactionMap.set(food, satisfaction);
        }

        const prefixSum = new Array(c.length + 1).fill(0);
        for (let i = 0; i < c.length; i++) {
            const satisfaction = satisfactionMap.get(c[i]) || 0.0;
            prefixSum[i + 1] = prefixSum[i] + satisfaction;
        }

        const result = new Array(q.length);
        for (let i = 0; i < q.length; i++) {
            const [start, end] = q[i];
            result[i] = prefixSum[end + 1] - prefixSum[start];
        }

        return result;
    }
}

module.exports = Judge;